### DISCUSSION TITLE
Source of CN / SAN in Azure OIDC cert

### DISCUSSION THREAD

**Discussion by dgitw** [smallstep/certificates#875] [Q&A] [answered]
I've generated a personal certificate using the [OIDC](https://smallstep.com/docs/certificate-manager/oidc) instructions with Azure as the provisioner. This works great!

My question is what is the source of the CN? It appears to be a token string?

The SAN also contains the string:
```
RFC822 Name=name@domain.com
URL=https://login.microsoftonline.com/[Tenant Id]/v2.0#[CN which looks like a token string]
```
and how can I change it?

**Comment by tashian** [comment]
Hi @dgitw, I'm guessing that URL is coming straight out of the OIDC token; it's not something our software would construct.
It may be the audience (`aud`) value.
Take a look at Microsoft's description of the fields of its OIDC token, in their OIDC integration docs.

**Comment by maraino** [comment, marked as answer]
@dgitw by default the OIDC provisioners sets the common name and SANs looking at token attributes:
* `sub` goes to the common name. It's quite common that `sub` matches to a user id.
* `email` becomes an email SAN
* `iss` if it can be parsed as a URI it goes to a URI SAN.

All this can be configured with a custom template. These templates have access to information in the token, the CSR, user-provided data, as well as the defaults that we set in the subject and sans. The templates docs are available at https://smallstep.com/docs/step-ca/templates/#x509-templates

**Reply by maraino** [reply]
OIDC provisioner is intended to identify humans, it can identify machines too, but it is not intended for that.

**Comment by dgitw** [comment]
Thanks @maraino - pointed me in right direction.

I ended up finding the token reference for Azure AD - https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens

`sub`  "The principal about which the token asserts information, such as the user of an app. This value is immutable and cannot be reassigned or reused. The subject is a pairwise identifier - it is unique to a particular application ID. If a single user signs into two different apps using two different client IDs, those apps will receive two different values for the subject claim. This may or may not be wanted depending on your architecture and privacy requirements."

Will try using templates to pull out something useful.
